src/node_report.cc

```diff
@@ -0,0 +1,1115 @@
+#include <Lm.h>
+#include <tchar.h>
+#include <psapi.h>
+#else
+#include <sys/resource.h>
+// Get the standard printf format macros for C99 stdint types.
+#ifndef __STDC_FORMAT_MACROS
+#define __STDC_FORMAT_MACROS
+#endif
+#include <inttypes.h>
+#include <cxxabi.h>
+#include <dlfcn.h>
+#if defined(__linux__) || defined(__sun)
+#include <link.h>
+#endif
+#ifdef _AIX
+#include <sys/ldr.h>  // ld_info structure
+#endif
+// Include execinfo.h for the native backtrace API. The API is
+// unavailable on AIX and on some Linux distributions, e.g. Alpine Linux.
+#if !defined(_AIX) && !(defined(__linux__) && !defined(__GLIBC__))
+#include <execinfo.h>
+#endif
+#include <sys/utsname.h>
+#endif
+
+#ifdef __APPLE__
+#include <mach-o/dyld.h>  // _dyld_get_image_name()
+#endif
+
+
+
+#ifndef _WIN32
+extern char** environ;
+#endif
+
+namespace nodereport {
+
+using v8::Exception;
+using v8::HeapSpaceStatistics;
+using v8::HeapStatistics;
+using v8::Isolate;
+using v8::Local;
+using v8::Message;
+using v8::RegisterState;
+using v8::SampleInfo;
+using v8::String;
+using v8::V8;
+
+// Internal/static function declarations
+static void WriteNodeReport(Isolate* isolate, DumpEvent event,
+                            const char* message, const char* location,
+                            char* filename, std::ostream &out,
+                            MaybeLocal<Value> error, TIME_TYPE* time);
+static void PrintCommandLine(std::ostream& out);
+static void PrintVersionInformation(std::ostream& out);
+static void PrintJavaScriptStack(std::ostream& out, Isolate* isolate,
+                                 DumpEvent event, const char* location);
+static void PrintJavaScriptErrorStack(std::ostream& out, Isolate* isolate,
+                                      MaybeLocal<Value> error);
+static void PrintStackFromStackTrace(std::ostream& out, Isolate* isolate,
+                                     DumpEvent event);
+static void PrintStackFrame(std::ostream& out, Isolate* isolate,
+                            Local<StackFrame> frame, int index, void* pc);
+static void PrintNativeStack(std::ostream& out);
+#ifndef _WIN32
+static void PrintResourceUsage(std::ostream& out);
+#endif
+static void PrintGCStatistics(std::ostream& out, Isolate* isolate);
+static void PrintSystemInformation(std::ostream& out, Isolate* isolate);
+static void PrintLoadedLibraries(std::ostream& out, Isolate* isolate);
+
+// Global variables
+static int seq = 0;  // sequence number for report filenames
+const char* v8_states[] = {"JS", "GC", "COMPILER", "OTHER", "EXTERNAL", "IDLE"};
+static bool report_active = false;  // recursion protection
+char report_filename[NR_MAXNAME + 1] = "";
+char report_directory[NR_MAXPATH + 1] = "";  // default: current directory
+TIME_TYPE loadtime_tm_struct;  // module load time
+time_t load_time;  // module load time absolute
+
+// std::string version_string = NODE_VERSION_STRING;
+// std::string commandline_string = "";
+version_and_command_struct version_and_command;
```

Comment:
This doesn’t seem ideal…

Is this supposed to reflect the flags under which the current Node.js instance (not process) operates, or should it be a snapshot of the actual command line flags as they were passed?

If it’s the latter, this should probably be handled as part of `node::Init`…

Reply:
yes, later - the original behavior is to extract the input the program as it is (and I guess that is the intent as well).

So moved the `commandline_string` logic out of here, intercepted it in `Node::Start` where the string is readily available as well as the Environment and Per-isolate data has been created and available for storing.